base/applications/cmdutils/tasklist/tasklist.c

```diff
@@ -0,0 +1,405 @@
+// Print memory size using appropriate unit, with comma-separated number, aligned to right.
+// MaxWidth is the width for printing.
+// StartingUnit is the minimum memory unit used, 0 for Byte, 1 for KB, 2 for MB ...
+// return FALSE when failed to format.
+BOOL PrintMemory(SIZE_T MemorySizeByte, INT MaxWidth, INT StartingUnit)
+{
+    WCHAR *MemoryUnit[] = {L"B", L"K", L"M", L"G", L"T", L"P"};
+
+    MaxWidth -= 2; // a space and a unit will occupy 2 length.
+    if (MaxWidth <= 0)
+    {
+        return FALSE;
+    }
+
+    SIZE_T MemorySize = MemorySizeByte;
+
+    for (INT i = 0; i < _countof(MemoryUnit); i++)
+    {
+        if (i >= StartingUnit)
+        {
+            // calculate the length needed to print.
+            INT PrintLength = 0;
+            SIZE_T Tmp = MemorySize;
+            do
+            {
+                Tmp /= 10;
+                PrintLength += 1;
+            } while (Tmp);
+
+            if (PrintLength + (PrintLength - 1) / 3 <= MaxWidth) // (PrintLength - 1) / 3 is the length comma will take.
+            {
+                // enough to hold
+
+                // print padding space
+                PrintSpace(MaxWidth - (PrintLength + (PrintLength - 1) / 3));
+
+                INT Mod = 1;
+                for (INT i = 0; i < PrintLength - 1; i++)
+                {
+                    Mod *= 10;
+                }
+
+                for (INT i = PrintLength - 1; i >= 0; i--)
+                {
+                    ConPrintf(StdOut, L"%d", MemorySize / Mod);
+                    MemorySize %= Mod;
+                    if (i && i % 3 == 0)
+                    {
+                        ConPrintf(StdOut, L",");
+                    }
+                    Mod /= 10;
+                }
+
+                // print the unit.
+                ConPrintf(StdOut, L" %ls", MemoryUnit[i]);
+
+                return TRUE;
+            }
+        }
+        
+        MemorySize >>= 10; // MemorySize /= 1024;
+    }
+
+    // out of MemoryUnit range.
+    return FALSE;
+}
+
+BOOL EnumProcessAndPrint(BOOL bVerbose)
+{
+    // Load ntdll.dll in order to use NtQuerySystemInformation
+    HMODULE hNtDLL = LoadLibraryW(L"Ntdll.dll");
+    if (!hNtDLL)
+    {
+        ConResMsgPrintf(StdOut, 0, IDS_ENUM_FAILED);
+        return FALSE;
+    }
+
+    NT_QUERY_SYSTEM_INFORMATION PtrNtQuerySystemInformation =
+        (NT_QUERY_SYSTEM_INFORMATION)GetProcAddress(hNtDLL, "NtQuerySystemInformation");
+
+    if (!PtrNtQuerySystemInformation)
+    {
+        ConResMsgPrintf(StdOut, 0, IDS_ENUM_FAILED);
+        FreeLibrary(hNtDLL);
+        return FALSE;
+    }
+
+    // Call NtQuerySystemInformation for the process information
+    ULONG ProcessInfoBufferLength = 0;
+    ULONG ResultLength = 0;
+    PBYTE ProcessInfoBuffer = 0;
+
+    // Get the buffer size we need
+    NTSTATUS Status = PtrNtQuerySystemInformation(SystemProcessInformation, NULL, 0, &ResultLength);
+
+    for (INT Retry = 0; Retry < NT_SYSTEM_QUERY_MAX_RETRY; Retry++)
```

Comment:
Following our previous discussion about that max_retry thing, could you add a comment that the purpose of this loop is to "mitigate" the case where new processes could appear between the NtQuerySystemInformation() call for size and the call for the actual data?

Reply:
OK, comment added! 👌